File: src/cli-launcher.ts
import { existsSync, realpathSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { basename, dirname, join, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";

export interface AimuxCliLaunchCommand {
  command: string;
  args: string[];
  source: "stable-shim" | "current-entry" | "native-binary";
  currentEntryPath: string;
  stableShimPath: string;
}

function fileExists(path: string): boolean {
  try {
    return existsSync(path) && statSync(path).isFile();
  } catch {
    return false;
  }
}

function currentEntryPath(): string {
  const compiledPath = fileURLToPath(new URL("./launcher-bin.js", import.meta.url));
  if (fileExists(compiledPath)) return compiledPath;
  return fileURLToPath(new URL("./launcher-bin.ts", import.meta.url));
}

function normalizeDir(path: string): string {
  const normalized = canonicalPath(path);
  return normalized.endsWith(sep) ? normalized : `${normalized}${sep}`;
}

function canonicalPath(path: string): string {
  try {
    return realpathSync(path);
  } catch {
    return resolve(path);
  }
}

function platformNativeBinaryPath(installRoot: string): string {
  return join(installRoot, "native", `${process.platform}-${process.arch}`, "aimux");
}

function nativeInstallRootFromEntry(path: string | undefined): string | null {
  const current = path?.trim();
  if (!current) return null;
  const canonical = canonicalPath(current);
  for (const suffix of [
    `${sep}dist${sep}launcher-bin.js`,
    `${sep}dist${sep}launcher-bin.ts`,
    `${sep}dist${sep}main.js`,
    `${sep}dist${sep}main.ts`,
    `${sep}bin${sep}aimux`,
  ]) {
    if (canonical.endsWith(suffix)) return canonical.slice(0, -suffix.length);
  }
  if (basename(canonical) === "aimux" && basename(dirname(canonical)) === `${process.platform}-${process.arch}`) {
    const nativeDir = dirname(dirname(canonical));
    if (basename(nativeDir) === "native") return dirname(nativeDir);
  }
  return null;
}

function nativeBinaryForStableShim(stableShimPath: string): string | null {
  try {
    const realShimPath = realpathSync(stableShimPath);
    if (basename(realShimPath) !== "aimux" || basename(dirname(realShimPath)) !== "bin") return null;
    const nativeBinary = platformNativeBinaryPath(dirname(dirname(realShimPath)));
    return fileExists(nativeBinary) ? nativeBinary : null;
  } catch {
    return null;
  }
}

function resolveInstalledNativeBinary(input: {
  currentArgvEntry: string | undefined;
  currentEntryPath: string;
  stableShimPath: string;
  env: NodeJS.ProcessEnv;
}): string | null {
  const explicit = input.env.AIMUX_NATIVE_BIN?.trim();
  if (explicit && fileExists(explicit)) return canonicalPath(explicit);
  const stableNative = nativeBinaryForStableShim(input.stableShimPath);
  if (stableNative && shouldUseStableShim(input)) return stableNative;
  for (const entry of [input.currentArgvEntry, input.currentEntryPath]) {
    const root = nativeInstallRootFromEntry(entry);
    if (!root) continue;
    const nativeBinary = platformNativeBinaryPath(root);
    if (fileExists(nativeBinary)) return nativeBinary;
  }
  return null;
}

export function getAimuxStableShimPath(env: NodeJS.ProcessEnv = process.env): string {
  return env.AIMUX_CLI_BIN?.trim() || `${homedir()}/.local/bin/aimux`;
}

function shouldUseStableShim(input: {
  currentArgvEntry: string | undefined;
  stableShimPath: string;
  env: NodeJS.ProcessEnv;
}): boolean {
  if (!fileExists(input.stableShimPath)) return false;
  const current = input.currentArgvEntry?.trim();
  if (!current) return false;
  if (canonicalPath(current) === canonicalPath(input.stableShimPath)) return true;
  const nativeRoot = normalizeDir(input.env.AIMUX_INSTALL_ROOT || `${homedir()}/.aimux/native`);
  return canonicalPath(current).startsWith(nativeRoot);
}

function resolveAimuxCliLaunchCommand(
  args: string[] = [],
  options: { env?: NodeJS.ProcessEnv; currentArgvEntry?: string; preferNativeBinary?: boolean } = {},
): AimuxCliLaunchCommand {
  const env = options.env ?? process.env;
  const stableShimPath = getAimuxStableShimPath(env);
  const currentEntry = currentEntryPath();
  const currentArgvEntry = options.currentArgvEntry ?? process.argv[1];
  if (options.preferNativeBinary) {
    const nativeBinary = resolveInstalledNativeBinary({
      currentArgvEntry,
      currentEntryPath: currentEntry,
      stableShimPath,
      env,
    });
    if (nativeBinary) {
      return {
        command: nativeBinary,
        args,
        source: "native-binary",
        currentEntryPath: nativeBinary,
        stableShimPath,
      };
    }
  }
  if (
    shouldUseStableShim({
      currentArgvEntry,
      stableShimPath,
      env,
    })
  ) {
    return {
      command: stableShimPath,
      args,
      source: "stable-shim",
      currentEntryPath: currentEntry,
      stableShimPath,
    };
  }
  return {
    command: process.execPath,
    args: [currentEntry, ...args],
    source: "current-entry",
    currentEntryPath: currentEntry,
    stableShimPath,
  };
}

export function getAimuxDaemonLaunchCommand(
  options: { env?: NodeJS.ProcessEnv; currentArgvEntry?: string } = {},
): AimuxCliLaunchCommand {
  return resolveAimuxCliLaunchCommand(["daemon", "run"], { ...options, preferNativeBinary: true });
}

export function getAimuxDashboardLaunchCommand(
  options: { env?: NodeJS.ProcessEnv; currentArgvEntry?: string } = {},
): AimuxCliLaunchCommand {
  if ((options.env ?? process.env).AIMUX_DASHBOARD_IMPLEMENTATION?.trim() === "native") {
    return resolveAimuxCliLaunchCommand(["__dashboard-internal-native"], { ...options, preferNativeBinary: true });
  }
  return resolveAimuxCliLaunchCommand(["--tmux-dashboard-internal"], options);
}

export function getAimuxProjectServiceLaunchCommand(
  projectId: string,
  projectRoot: string,
  options: { env?: NodeJS.ProcessEnv; currentArgvEntry?: string } = {},
): AimuxCliLaunchCommand {
  return resolveAimuxCliLaunchCommand(
    ["__project-service-internal", "--project-id", projectId, "--project-root", projectRoot],
    { ...options, preferNativeBinary: true },
  );
}

export function getAimuxCurrentCliIdentity(
  options: { env?: NodeJS.ProcessEnv; currentArgvEntry?: string } = {},
): AimuxCliLaunchCommand {
  return resolveAimuxCliLaunchCommand([], { ...options, preferNativeBinary: true });
}

File: native/crates/aimux/src/cli_launcher.rs
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AimuxCliLaunchSource {
    StableShim,
    CurrentEntry,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AimuxCliLaunchCommand {
    pub command: String,
    pub args: Vec<String>,
    pub source: AimuxCliLaunchSource,
    pub current_entry_path: String,
    pub stable_shim_path: String,
}

#[derive(Debug, Clone, Default)]
pub struct AimuxCliLaunchOptions {
    pub env: BTreeMap<String, String>,
    pub current_argv_entry: Option<String>,
    pub current_entry_path: Option<String>,
    pub home_dir: Option<PathBuf>,
}

pub fn get_aimux_stable_shim_path() -> String {
    get_aimux_stable_shim_path_from(&std_env(), home_dir())
}

pub fn get_aimux_stable_shim_path_from(
    env: &BTreeMap<String, String>,
    home_dir: PathBuf,
) -> String {
    env.get("AIMUX_CLI_BIN")
        .map(|value| value.trim())
        .filter(|value| !value.is_empty())
        .map(str::to_owned)
        .unwrap_or_else(|| {
            home_dir
                .join(".local/bin/aimux")
                .to_string_lossy()
                .into_owned()
        })
}

pub fn get_aimux_daemon_launch_command(options: AimuxCliLaunchOptions) -> AimuxCliLaunchCommand {
    resolve_aimux_cli_launch_command(vec!["daemon".into(), "run".into()], options)
}

pub fn get_aimux_dashboard_launch_command(options: AimuxCliLaunchOptions) -> AimuxCliLaunchCommand {
    let args = dashboard_launch_args(&options.env);
    resolve_aimux_cli_launch_command(args, options)
}

pub fn get_aimux_project_service_launch_command(
    project_id: &str,
    project_root: &str,
    options: AimuxCliLaunchOptions,
) -> AimuxCliLaunchCommand {
    resolve_current_entry_launch_command(
        project_service_launch_args(project_id, project_root),
        options,
    )
}

pub fn get_aimux_current_cli_identity(options: AimuxCliLaunchOptions) -> AimuxCliLaunchCommand {
    resolve_aimux_cli_launch_command(Vec::new(), options)
}

pub fn resolve_aimux_cli_launch_command(
    args: Vec<String>,
    options: AimuxCliLaunchOptions,
) -> AimuxCliLaunchCommand {
    let env = options.env;
    let home_dir = options.home_dir.unwrap_or_else(home_dir);
    let stable_shim_path = get_aimux_stable_shim_path_from(&env, home_dir.clone());
    let current_entry_path = options
        .current_entry_path
        .unwrap_or_else(current_entry_path);
    if should_use_stable_shim(&ShouldUseStableShimInput {
        current_argv_entry: options.current_argv_entry.as_deref(),
        stable_shim_path: &stable_shim_path,
        env: &env,
        home_dir: &home_dir,
    }) {
        return AimuxCliLaunchCommand {
            command: stable_shim_path.clone(),
            args,
            source: AimuxCliLaunchSource::StableShim,
            current_entry_path,
            stable_shim_path,
        };
    }
    AimuxCliLaunchCommand {
        command: current_entry_path.clone(),
        args,
        source: AimuxCliLaunchSource::CurrentEntry,
        current_entry_path,
        stable_shim_path,
    }
}

fn resolve_current_entry_launch_command(
    args: Vec<String>,
    options: AimuxCliLaunchOptions,
) -> AimuxCliLaunchCommand {
    let env = options.env;
    let home_dir = options.home_dir.unwrap_or_else(home_dir);
    let stable_shim_path = get_aimux_stable_shim_path_from(&env, home_dir);
    let current_entry_path = options
        .current_entry_path
        .unwrap_or_else(current_entry_path);
    AimuxCliLaunchCommand {
        command: current_entry_path.clone(),
        args,
        source: AimuxCliLaunchSource::CurrentEntry,
        current_entry_path,
        stable_shim_path,
    }
}

fn project_service_launch_args(project_id: &str, project_root: &str) -> Vec<String> {
    vec![
        "__project-service-internal".into(),
        "--project-id".into(),
        project_id.into(),
        "--project-root".into(),
        project_root.into(),
    ]
}

fn dashboard_launch_args(env: &BTreeMap<String, String>) -> Vec<String> {
    if env
        .get("AIMUX_DASHBOARD_IMPLEMENTATION")
        .map(|value| value.trim())
        == Some("node")
    {
        vec!["--tmux-dashboard-internal".into()]
    } else {
        vec!["__dashboard-internal-native".into()]
    }
}

struct ShouldUseStableShimInput<'a> {
    current_argv_entry: Option<&'a str>,
    stable_shim_path: &'a str,
    env: &'a BTreeMap<String, String>,
    home_dir: &'a Path,
}

fn should_use_stable_shim(input: &ShouldUseStableShimInput<'_>) -> bool {
    if !file_exists(input.stable_shim_path) {
        return false;
    }
    let Some(current) = input
        .current_argv_entry
        .map(str::trim)
        .filter(|value| !value.is_empty())
    else {
        return false;
    };
    if canonical_path(current) == canonical_path(input.stable_shim_path) {
        return true;
    }
    let native_root = input
        .env
        .get("AIMUX_INSTALL_ROOT")
        .map(|value| value.as_str())
        .filter(|value| !value.trim().is_empty())
        .map(PathBuf::from)
        .unwrap_or_else(|| input.home_dir.join(".aimux/native"));
    canonical_path(current).starts_with(&normalize_dir(native_root))
}

fn file_exists(path: impl AsRef<Path>) -> bool {
    std::fs::metadata(path)
        .map(|metadata| metadata.is_file())
        .unwrap_or(false)
}

fn current_entry_path() -> String {
    std::env::current_exe()
        .or_else(|_| std::env::current_dir().map(|cwd| cwd.join("aimux")))
        .unwrap_or_else(|_| PathBuf::from("aimux"))
        .to_string_lossy()
        .into_owned()
}

fn normalize_dir(path: impl AsRef<Path>) -> String {
    let mut normalized = canonical_path(path);
    let separator = std::path::MAIN_SEPARATOR;
    if !normalized.ends_with(separator) {
        normalized.push(separator);
    }
    normalized
}

fn canonical_path(path: impl AsRef<Path>) -> String {
    let path = path.as_ref();
    std::fs::canonicalize(path)
        .unwrap_or_else(|_| {
            if path.is_absolute() {
                path.to_path_buf()
            } else {
                std::env::current_dir()
                    .unwrap_or_else(|_| PathBuf::from("."))
                    .join(path)
            }
        })
        .to_string_lossy()
        .into_owned()
}

fn home_dir() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}

fn std_env() -> BTreeMap<String, String> {
    std::env::vars().collect()
}
